### PR TITLE
add __deepcopy__ back to Parameter

### LINKED ISSUE #315: copy of Parameter converts it to Variable
```
In [27]: l = nn.Linear(10,1)

In [28]: s = l.state_dict()

In [29]: s2 = copy.deepcopy(s)

In [32]: s
Out[32]:
OrderedDict([('bias', Parameter containing:
              -0.1027
              [torch.FloatTensor of size 1]), ('weight', Parameter containing:
               0.1287  0.2447  0.0268  0.1237  0.2291 -0.2262  0.2658 -0.1941 -0.2136  0.0907
              [torch.FloatTensor of size 1x10])])
In [36]: s2
Out[36]:
OrderedDict([('bias', Variable containing:
              -0.1027
              [torch.FloatTensor of size 1]), ('weight', Variable containing:
               0.1287  0.2447  0.0268  0.1237  0.2291 -0.2262  0.2658 -0.1941 -0.2136  0.0907
              [torch.FloatTensor of size 1x10])])
```

### PR DESCRIPTION
**Pull request by weiyangfb**
- fix https://github.com/pytorch/pytorch/issues/315
- add `__deepcopy__` back to Parameter class

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -7370,6 +7370,17 @@ def test_deepcopy_scalar(self):
...
         self.assertEqual(a, deepcopy(a))
 
+    def test_deepcopy_parameter(self):
+        from copy import deepcopy
+        l = torch.nn.Linear(10, 1)
+        s = l.state_dict(keep_vars=True)
+        self.assertEqual(torch.nn.Parameter, type(s['weight']))
+        self.assertEqual(torch.nn.Parameter, type(s['bias']))
+
+        s2 = copy.deepcopy(s)
```
Comment (vishwakftw) [inline test/test_torch.py]: nit: just deepcopy

File: torch/nn/parameter.py [hunk 2]
```diff
@@ -24,6 +24,11 @@ def __new__(cls, data=None, requires_grad=True):
             data = torch.Tensor()
         return torch.Tensor._make_subclass(cls, data, requires_grad)
 
+    def __deepcopy__(self, memo):
+        result = type(self)(self.data.clone(), self.requires_grad)
```
Comment (vishwakftw) [inline torch/nn/parameter.py]: Can't we do this conditionally?
```suggestion
        if id(self) not in memo:
            # do this
        else:
            return memo[id(self)]
```

Comment (weiyangfb) [reply torch/nn/parameter.py]: Thanks! Yeah, I forgot to do that.

**Comment by weiyangfb** [comment]
is this ready to go? @apaszke @soumith

**Review by apaszke — APPROVED** [review]

**Pull request closed without merging**